open_ce/utils.py

```diff
@@ -226,6 +226,40 @@ def cuda_driver_installed():
 
         raise OpenCEError(Error.ERROR, "lsmod command unexpectedly failed") from err
 
+def check_cuda_version_match(command):
+    '''
+    Check whether cudatoolkit matches the system CUDA version (as per CUDA_HOME).
+    If not, print a warning.  Returns True/False based on match.
+    '''
+
+    # Version is recorded in $CUDA_HOME/version.txt; make sure $CUDA_HOME is set first.
+    # (If $CUDA_HOME doesn't exist, read will fail so we return False).
+    check_cuda_home()
+    cudavers_file = os.environ['CUDA_HOME'] + '/version.txt'
+    try:
+        cudafile = open(cudavers_file,'r')
+        cuda_home_version = cudafile.readline()
+        version_match = cuda_home_version.find(str(command.cudatoolkit))
+        cudafile.close()
```

Comment:
Should this line be put into finally block?

Reply:
I could change this - do you think that is better?
In this simple usage case, since I was opening the file only to read one line (to read the version) and am then done with
the file, I thought I would close it immediately after the use.  (If there was an open error, the 'except' will
catch that). So I am not sure if it makes much difference in this case if I use a finally or not....  unless that is simply
desirable as better python coding practice  (?)

Reply:
Yes, I think it is always better to release resources in the finally block as it will be always executed. But in your case, if it is just a single line being read from a file during which there is hardly a chance of any mishap, you can skip this comment. As a better python practice, indeed it should go to finally block.

Reply:
Thanks. I'm thinking: even if it seems not to be significant in this case, since I am not an expert with the python language I should be as strict as I can about following best practices, so I will change this as per this advice.

Reply:
@npanpaliya After making the suggested change, the code bombs if the file open() had failed (i.e. if the file wasn't there). If the open fails, the exception message is printed and then the 'finally' with the close() is attempted, but this gives an error since the file isn't open (doesn't exist); we get an `UnboundLocalError` at that line.
In the previous form, the close is only attempted if we are still in the 'try' block at that point, which means the open had succeeded.   So it seems that that I should put this back how it was.

Reply:
@dncliss - This is how I tried and it worked -
```
diff --git a/open_ce/utils.py b/open_ce/utils.py
index 85b3e05..197f01b 100644
--- a/open_ce/utils.py
+++ b/open_ce/utils.py
@@ -236,16 +236,19 @@ def check_cuda_version_match(command):
     # (If $CUDA_HOME doesn't exist, read will fail so we return False).
     check_cuda_home()
     cudavers_file = os.environ['CUDA_HOME'] + '/version.txt'
+    cudafile = None
     try:
         cudafile = open(cudavers_file,'r')
         cuda_home_version = cudafile.readline()
-        cudafile.close()
         version_match = cuda_home_version.find(str(command.cudatoolkit))
         if version_match > 0:
             return True
     except Exception:  # pylint: disable=broad-except
         # Treating as a warning rather than a fatal error
         print("WARNING: Could not read version from " + cudavers_file)
+    finally:
+        if cudafile:
+            cudafile.close()
```

Reply:
@npanpaliya Ah, that makes complete sense; it needs to be initialized and checked so the close() won't be attempted if the open() had failed.  I'll correct and push this part again.